test/jest/Bladeburner/Actions.test.ts

```diff
@@ -0,0 +1,358 @@
+        expect(inst.rank).toBeLessThan(before);
+      });
+    });
+  });
+
+  it("have a minimum duration of 1 second", () => {
+    complete(SampleContract);
+    expect(inst.actionTimeToComplete).toBeGreaterThanOrEqual(1);
+  });
+
+  beforeAll(() => {
+    /* Initialise Formatters. Dependency of Bladeburner Logs/Console */
+    FormatsNeedToChange.emit();
+  });
+
+  beforeEach(() => {
+    setPlayer(new PlayerObject());
+
+    /** Need BN5 to receive Int EXP */
+    Player.sourceFiles.set(5, 3);
+
+    if (initBladeburner(Player)) {
+      inst = Player.bladeburner;
+      inst.clearConsole();
+    }
+  });
+
+  function initBladeburner(player: PlayerObject): player is PlayerObject & { bladeburner: Bladeburner } {
+    player.startBladeburner();
+    return true;
+  }
+
+  function basicStats() {
+    inst.rank = 1;
+    inst.changeRank(Player, 10);
+    Player.gainStrengthExp(LOW_STAT_EXP);
+    Player.gainDefenseExp(LOW_STAT_EXP);
+    Player.gainAgilityExp(LOW_STAT_EXP);
+    Player.gainDexterityExp(LOW_STAT_EXP);
+
+    inst.calculateMaxStamina();
+    inst.stamina = inst.maxStamina;
+
+    resetCity();
+  }
+
+  function guaranteeSuccess() {
+    inst.rank = 1;
+    inst.changeRank(Player, 1e10);
+
+    Player.gainStrengthExp(HIGH_STAT_EXP);
+    Player.gainDefenseExp(HIGH_STAT_EXP);
+    Player.gainAgilityExp(HIGH_STAT_EXP);
+    Player.gainDexterityExp(HIGH_STAT_EXP);
+    inst.setSkillLevel(BladeburnerSkillName.BladesIntuition, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.Cloak, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.DigitalObserver, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.Reaper, 1e12);
+
+    inst.calculateMaxStamina();
+    inst.stamina = inst.maxStamina;
+
+    resetCity();
+  }
+
+  function guaranteeFailure() {
+    inst.rank = 1;
+    inst.changeRank(Player, 400e3); // Minimum to attempt to fail hardest op
+    Player.gainAgilityExp(1e50);
+    inst.calculateMaxStamina();
+    inst.stamina = 0;
+
+    resetCity();
+  }
+
+  function resetCity() {
+    inst.cities[inst.city].chaos = 0;
+    inst.cities[inst.city].comms = 100;
+    inst.cities[inst.city].pop = 1e9;
+
+    /** Disable random event */
+    inst.randomEventCounter = Infinity;
+  }
+
+  function allCitiesHighChaos() {
+    for (const city of Object.values(inst.cities)) {
+      city.chaos = 1e12;
+    }
+  }
+
+  function complete(id: ActionIdentifier) {
+    start(id), finish();
+  }
+
+  function start(id: ActionIdentifier) {
+    const action = inst.getActionObject(id);
+    if ("count" in action) action.count = 1;
+    if (action.type === BladeburnerActionType.Operation) action.autoLevel = true;
+    if (id.type === "Black Operations") inst.numBlackOpsComplete = (<BlackOperation>action).n;
+    inst.startAction(id);
+  }
+
+  function finish() {
+    inst.processAction(ENOUGH_TIME_TO_FINISH_ACTION);
+    inst.calculateMaxStamina();
+  }
+
+  function* nonGeneralActions() {
```

Comment:
As an example of what I'm talking about below, as a non-generator this is as simple as:
`return [bb.contracts, bb.operations, BlackOperations].flatMap(Object.values);`